crate-tmp/src/nvtxw.rs

```diff
@@ -0,0 +1,248 @@
+pub const NVTX_PAYLOAD_SCHEMA_TYPE_STATIC: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_TYPE_STATIC as u64;
+pub const NVTX_PAYLOAD_SCHEMA_TYPE_DYNAMIC: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_TYPE_DYNAMIC as u64;
+pub const NVTX_PAYLOAD_SCHEMA_FLAG_NONE: u64 = nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_FLAG_NONE as u64;
+pub const NVTX_PAYLOAD_SCHEMA_FLAG_REFERENCED: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_FLAG_REFERENCED as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_NAME: u64 = nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_NAME as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_TYPE: u64 = nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_TYPE as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_FLAGS: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_FLAGS as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_ENTRIES: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_ENTRIES as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_NUM_ENTRIES: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_NUM_ENTRIES as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_STATIC_SIZE: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_STATIC_SIZE as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_SCHEMA_ID: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_SCHEMA_ID as u64;
+
+pub const NVTXW3_STREAM_ORDER_INTERLEAVING_NONE: i16 = 0;
+pub const NVTXW3_STREAM_ORDERING_TYPE_UNKNOWN: i16 = 0;
+
+fn check(result: ResultCode) -> Result<(), ResultCode> {
+    if result != NVTXW3_RESULT_SUCCESS {
+        return Err(result);
+    }
+    Ok(())
+}
+
+pub fn initialize_simple() -> Result<InterfaceHandle, ResultCode> {
+    let so_name = "libNvtxwBackend.so";
+    let c_mode_string = CString::new(so_name).expect("modeString CString::new failed");
+    let cptr_mode_string = c_mode_string.as_ptr();
+
+    let mut get_interface_func: nvtxw_bindings::nvtxwGetInterface_t = Default::default();
+    check(unsafe {
+        nvtxw_bindings::nvtxwInitialize(
+            NVTXW3_INIT_MODE_LIBRARY_FILENAME,
+            cptr_mode_string,
+            &mut get_interface_func,
+            null_mut(),
+        )
+    })?;
+
+    let get_interface = get_interface_func.ok_or(NVTXW3_RESULT_FAILED)?;
+
+    let mut interface_void: *const c_void = null();
+    check(unsafe { get_interface(NVTXW3_INTERFACE_ID_CORE_V1, &mut interface_void) })?;
+
+    if interface_void.is_null() {
+        return Err(NVTXW3_RESULT_FAILED);
+    }
+
+    let ptr_interface = interface_void as *const nvtxw_bindings::nvtxwInterfaceCore_t;
+    let interface;
+    unsafe {
+        interface = *ptr_interface;
+    }
+
+    Ok(interface)
+}
+
+pub fn session_begin_simple(
+    interface: InterfaceHandle,
+    o_output: Option<PathBuf>,
+    o_merge: Option<PathBuf>,
+) -> Result<SessionHandle, ResultCode> {
+    let func = interface
+        .SessionBegin
+        .ok_or(NVTXW3_RESULT_INVALID_ARGUMENT)?;
+
+    let mut c_output: CString = CString::new("").expect("CString::new failed");
+    if let Some(os_output) = o_output {
```

Comment:
Ok, there's a bunch of stuff to say about this:

 1. This code double-allocates the `CString` in the case where `o_output` is `None`
 2. The code uses an empty `CString` as an error case which is weird and non-idiomatic
 3. The logic itself still uses direct string manipulation
 4. The line `CString::new(...).expect(...)` gets repeated all over the place

If I understand what you're trying to do, you'd get a very similar logic with:

```rust
let c_output = o_output.and_then(|o| Path::new(o).file_stem()).map(|s| CString::new(s).expect("output contains nul bytes"));
```

Note that this strips *any* file extension rather than just `.nsys-rep`. If you feel very strongly about that specific behavior (and that if a user who uses a non-default extension should *not* have it stripped), then we'd have to tweak the approach.

Since you use the empty case as an error anyway, it doesn't seem worth allocating a `CString` just to check its emptiness. Just let it be `None` on failure and then later on you can do:

```rust
c_output.map_or_else(null_mut, |o| o.as_ptr())
```

As a general principle, you do not want to use sentinels in Rust, and instead you want to directly encode error states so that when things go bad it's well-defined what happens.

Reply:
I've updated this part of the code. (It turns out `.as_ref()` is important in multiple places.) Please let me know if it can be improved further.